src/StreamJsonRpc/JsonRpc.cs

```diff
@@ -933,11 +940,23 @@ private JsonRpcMessage HandleInvocationTaskResult(JToken id, Task t)
                 throw new ArgumentException(Resources.TaskNotCompleted, nameof(t));
             }
 
-            if (t.IsFaulted)
+            if (t.IsFaulted && !this.ShouldCloseStreamOnException())
             {
                 return CreateError(id, t.Exception);
             }
 
+            if (t.IsFaulted && this.ShouldCloseStreamOnException())
+            {
+                if (t.Exception is AggregateException && t.Exception.InnerException != null)
+                {
+                    // Never let the outer (TargetInvocationException) escape because the inner is the interesting one to the caller, the outer is due to
+                    // the fact we are using reflection.
+                    throw t.Exception.InnerException;
+                }
+
+                throw t.Exception;
```

Comment:
Rethrowing an Exception like this stomps on its Stacktrace property with a new one, blowing away the evidence of the original one. Instead, rethrowing should be done with `ExceptionDispatchInfo.Capture(t).Throw()`.